README.md
# csDrummer
Csound Record a drum session in a virtual studio
First of all you need Csound 5.09 or late. Versions before 5.09 have a serious bug in morphing opcode. Software works but this results in a more synthetic sounds.

After you obtain Csound and choose MIDI file enter this in command-line: csound -o Output.wav -F Input.mid -T -d Drummer.orc Generated sound will be written to Output.wav and rendering will be terminated when the end of MIDI file is reached.

You may also use additional options: csound -o Output.wav -F Input.mid -T -d --omacro:Option=Value Drummer.orc

Available options: | Realtime | Set it to any value to use the low quality realtime mix-morphing (SVN version only) | |:---------|:------------------------------------------------------------------------------------| | Multichannel | Set it to any value to generate multichannel audio file. | | HumanAmp | Value of 0..1. Affects the hit force altering. | | HumanMorph | Value of 0..1. Affects the morphing altering. | | HumanTime | Value in sec. Affects the hit moment altering. | | HumanFalloff | Value in sec. Affects hit force falling off at high speed. (SVN version only) |

You can also refer to Csound command-line options. http://www.csounds.com/manual/html/CommandFlagsCategory.html


trunk/ChangeLog
SVN

    * console interface
    
    * rewritten engine
    
    * humanizing force falloff at high bpm
    
    * XML format for drumkits
    
    * unlimited number of velocities layers
    
    * low quality mode for realtime composing

20080926
    
    * morph between two velocities layers
    
    * humanizing time, hit force, morphing
    
    * multichannel/mono mode



trunk/src/test.py
import unittest
import random

import drumkit
##import csound
import sampler

class TestSampler(unittest.TestCase):
    
    def setUp(self):
        
        self.sampler = sampler.Sampler()
        
        # new instrument
        instr = self.sampler.createInstrument(0)
        # new technique
        tech = self.sampler.createTechnique(instr)
        self.sampler.setTechniqueMIDI(instr, tech, 35)
        self.sampler.setTechniqueMIDI(instr, tech, 36)
        # new target
        target = self.sampler.createTarget(instr, tech)
        # new layer
        layer = self.sampler.createLayer(instr, tech, target)
        layer = self.sampler.createLayer(instr, tech, target)        
        # new target
        target = self.sampler.createTarget(instr, tech)
        # new layer
        layer = self.sampler.createLayer(instr, tech, target)
        layer = self.sampler.createLayer(instr, tech, target)                
        layer = self.sampler.createLayer(instr, tech, target)                
        # sort
        self.sampler.sortLayersByVel(instr, tech, target)
        
        # new instrument
        instr = self.sampler.createInstrument(0)
        # new technique
        tech = self.sampler.createTechnique(instr)
        self.sampler.setTechniqueMIDI(instr, tech, 38)
        self.sampler.setTechniqueMIDI(instr, tech, 40)
        # new target
        target = self.sampler.createTarget(instr, tech)
        # new layer
        layer = self.sampler.createLayer(instr, tech, target)
        self.sampler.setLayerVelocity(instr, tech, target, layer, 0.33)
        layer = self.sampler.createLayer(instr, tech, target)        
        self.sampler.setLayerVelocity(instr, tech, target, layer, 0.75)
        layer = self.sampler.createLayer(instr, tech, target)        
        self.sampler.setLayerVelocity(instr, tech, target, layer, 0.25)        
        layer = self.sampler.createLayer(instr, tech, target)        
        self.sampler.setLayerVelocity(instr, tech, target, layer, 0.66)
        # sort
        self.sampler.sortLayersByVel(instr, tech, target)
        
        self.sampler.setSampleData(0, 100)
        self.sampler.setSampleData(1, 101)


    def testMapping(self):
        
        # check MIDI
       
        assert (int(self.sampler.isExists(35)) == 1)
        assert (int(self.sampler.isExists(36)) == 1)
        assert (int(self.sampler.isExists(38)) == 1)
        assert (int(self.sampler.isExists(40)) == 1)
        
        # check isntruments
        
        instr1 = int(self.sampler.getInstrument(35))
        instr2 = int(self.sampler.getInstrument(36))
        assert (instr1 == instr2)
        instr1 = int(self.sampler.getInstrument(35))
        instr2 = int(self.sampler.getInstrument(40))        
        assert (instr1 != instr2)
       
    
    def testSamplesData(self):
        
        sd = int(self.sampler.getSampleData(0))
        assert(sd == 100)
        sd = int(self.sampler.getSampleData(1))
        assert(sd == 101)
        
    
    def testHuman(self):
        
        # falloff
        
        vel = 0.66
        falloff = self.sampler.humanFalloff(vel, 0.33, 0)
        assert (falloff == vel)
        vel = 0.66
        falloff = self.sampler.humanFalloff(vel, 0.33, 0.5)
        assert (falloff < vel)        
        
        # force
        
        vel = 0.66
        force = self.sampler.humanVelocity(vel, 0.33)
        assert (force >= 0 and force <= 1)
        
        # time 
        
        human = 0.33
        time = self.sampler.humanTime(human)
        assert (time >= 0 and time <= human)
        
    
    def testMorph(self):
        
        midi = 40
        target = 0
        
        # check normal mode

        vel = 0.3
        top, bot, morph = self.sampler.getMorph(midi, target, vel)
        assert (morph >= 0 and morph <= 1)

        vel = 0.6
        top, bot, morph = self.sampler.getMorph(midi, target, vel)
        assert (morph >= 0 and morph <= 1)
        
        # check exceed bounds
        
        vel = 0.1
        top, bot, morph = self.sampler.getMorph(midi, target, vel)
        assert (morph >= 0 and morph <= 1)
        
        vel = 0.9
        top, bot, morph = self.sampler.getMorph(midi, target, vel)
        assert (morph >= 0 and morph <= 1)
        
        
class TestDrumkit(unittest.TestCase):
    
    testXML = """
<Drumkit>
<Version>1</Version>
<Description>test</Description>
<License></License>
<Instrument Name="Instrument1">
<Technique>
<MIDI>35</MIDI>
<MIDI>36</MIDI>
<Target>
<Channel>Mono</Channel>
<Gain>1</Gain>
<ADSR><Attack>0.01</Attack><Decay>0</Decay><Sustain>1</Sustain><Release>0.1</Release></ADSR>
<Layer><Velocity>1</Velocity><Gain>1</Gain><Sample>test.wav</Sample></Layer>
<Layer><Velocity>0</Velocity><Gain>1</Gain><Sample>test2.wav</Sample></Layer>
</Target>
</Technique>
</Instrument>
</Drumkit>
"""

    def testParser(self):
        
        testkit = drumkit.parseXML(self.testXML)
        channels = testkit.getChannels()
        assert (len(channels) == 1)
        assert ("Mono" in channels)
        samples = testkit.getSamples()
        assert (len(samples) == 2)
        assert ("test.wav" in samples)        
        assert ("test2.wav" in samples)                


if __name__ == '__main__':
    unittest.main()


branches/new-engine/test.py
import unittest

import drumkit
import csound
import sampler

class TestAll(unittest.TestCase):

    xmlfile = "DrumKit.xml"

    def test_drumkit(self):

        print "Parsing", self.xmlfile
        drumkit_parser = drumkit.DrumKitParser(self.xmlfile)
        print "drumkit_parser", drumkit_parser
        print "Channels", drumkit_parser.get_channels()
        print "Samples", drumkit_parser.get_samples()
        for instrument in drumkit_parser.get_instruments():
            print "\tInstrument", instrument
            for technique in instrument.get_techniques():
                print "\t\tTechnique", technique
                print "\t\tMIDI", technique.get_midi()
                for target in technique.get_targets():
                    print "\t\t\tTarget", target
                    print "\t\t\tChannel", target.get_channel()
                    print "\t\t\tADSR", target.get_adsr()
                    for layer in target.get_layers():
                        print "\t\t\t\tLayer", layer
                        print "\t\t\t\tVelocity", layer.get_velocity()
                        print "\t\t\t\tGain", layer.get_gain()
                        print "\t\t\t\tSample", layer.get_sample()


    def test_csdgen(self):

        drumkit_src = drumkit.DrumKitParser(self.xmlfile)
        drumkit_dest = csound.import_drumkit(drumkit_src)
        loader = drumkit_dest.get_text()
        sampler = csound.CsoundGenerator()
        sampler.drumkit_loader = loader
        code = sampler.get_text()
        print code


    def test_sampler(self):

        print "maybe should write test.csd?"


if __name__ == '__main__':
    unittest.main()


trunk/src/drumkit.py
# csDrummer: Record a drum session in a virtual studio
# Copyright (C) 2008 Schlagg
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import xml.dom.minidom


class LayerParser:
    """parser for <Layer> tag"""

    def __init__(self, layer):

        self.layer = layer


    def getVelocity(self):

        tag = self.layer.getElementsByTagName("Velocity")[0]
        assert tag.parentNode == self.layer
        vel = float(tag.firstChild.data)
        return vel


    def getGain(self):

        tag = self.layer.getElementsByTagName("Gain")[0]
        assert tag.parentNode == self.layer
        gain = float(tag.firstChild.data)
        return gain


    def getSample(self):

        tag = self.layer.getElementsByTagName("Sample")[0]
        assert tag.parentNode == self.layer
        sample = str(tag.firstChild.data)
        return sample


class TargetParser:
    """parser for <Target> tag"""

    def __init__(self, target):

        self.target = target


    def getChannel(self):

        tag = self.target.getElementsByTagName("Channel")[0]
        assert tag.parentNode == self.target
        channel = str(tag.firstChild.data)
        return channel


    def getGain(self):

        tag = self.target.getElementsByTagName("Gain")[0]
        assert tag.parentNode == self.target
        gain = float(tag.firstChild.data)
        return gain


    def getADSR(self):

        tag = self.target.getElementsByTagName("Attack")[0]
        assert tag.parentNode.parentNode == self.target
        attack = float(tag.firstChild.data)
        tag = self.target.getElementsByTagName("Decay")[0]
        assert tag.parentNode.parentNode == self.target
        decay = float(tag.firstChild.data)
        tag = self.target.getElementsByTagName("Sustain")[0]
        assert tag.parentNode.parentNode == self.target
        sustain = float(tag.firstChild.data)
        tag = self.target.getElementsByTagName("Release")[0]
        assert tag.parentNode.parentNode == self.target
        release = float(tag.firstChild.data)
        return (attack, decay, sustain, release)


    def getLayers(self):

        tags = self.target.getElementsByTagName("Layer")
        layers = []
        for tag in tags:
            assert tag.parentNode == self.target            
            layer = LayerParser(tag)
            layers += [layer]
        return layers


class TechniqueParser:
    """parser for <Technique> tag"""

    def __init__(self, technique):

        self.technique = technique
        
        
    def getName(self):
        
        name = str(self.technique.getAttribute("Name"))
        return name


    def getMIDI(self):

        tags = self.technique.getElementsByTagName("MIDI")
        notes = []
        for tag in tags:
            assert tag.parentNode == self.technique
            note = int(tag.firstChild.data)
            assert (note >= 0) and (note < 128)
            notes += [note]
        return notes


    def getTargets(self):

        tags = self.technique.getElementsByTagName("Target")
        targets = []
        for tag in tags:
            assert tag.parentNode == self.technique
            target = TargetParser(tag)
            targets += [target]
        return targets


class InstrumentParser:
    """parser for <Instrument> tag"""

    def __init__(self, instrument):

        self.instrument = instrument

    
    def getName(self):
        
        name = str(self.technique.getAttribute("Name"))
        return name
        

    def getTechniques(self):

        tags = self.instrument.getElementsByTagName("Technique")
        techniques = []
        for tag in tags:
            assert tag.parentNode == self.instrument
            technique = TechniqueParser(tag)
            techniques += [technique]
        return techniques


class DrumkitParser:
    """parser for entire drum kit"""

    def __init__(self, document):

        self.document = document
        assert self.document.documentElement.tagName == "Drumkit"


    def getChannels(self):

        tags = self.document.getElementsByTagName("Channel")
        channels = []
        for tag in tags:
            assert tag.parentNode.tagName == "Target"
            channel = str(tag.firstChild.data)
            channels += [channel]
        channels = list(set(channels))
        return channels


    def getSamples(self):

        tags = self.document.getElementsByTagName("Sample")
        samples = []
        for tag in tags:
            assert tag.parentNode.tagName == "Layer"
            sample = str(tag.firstChild.data)
            samples += [sample]
        samples = list(set(samples))
        return samples


    def getInstruments(self):

        tags = self.document.getElementsByTagName("Instrument")
        instruments = []
        for tag in tags:
            assert tag.parentNode == self.document.documentElement
            instrument = InstrumentParser(tag)
            instruments += [instrument]
        return instruments


def parseFile(filename):
    
    document = xml.dom.minidom.parse(filename)
    drumkit = DrumkitParser(document)
    return drumkit
    

def parseXML(text):
    
    document = xml.dom.minidom.parseString(text)
    drumkit = DrumkitParser(document)
    return drumkit

branches/new-engine/csdrummer.py
import optparse
import sys
import time

from csound import *



class CommandLineParser:

    def __init__(self):

        # create parser
        self.parser = optparse.OptionParser(
            usage="%prog [options]")
        # add options
        self.parser.add_option(
            "--human-force", metavar="VALUE",
            dest="HumanForce", default=0.25,
            help="hit force altering (0..1), default 0.25")
        self.parser.add_option(
            "--human-time", metavar="VALUE",
            dest="HumanTime", default=0.033,
            help="hit moment altering (sec), default 0.033")
        self.parser.add_option(
            "--human-falloff", metavar="VALUE",
            dest="HumanFalloff", default=0.33,
            help="hit force falling off (sec), default 0.33")
        input_modules = str(InputModuleGenerator.modules.keys())
        self.parser.add_option(
            "--input-module", metavar="MODULE",
            dest="InputModule", default="portmidi",
            help="input modules: " + input_modules)
        output_modules = str(OutputModuleGenerator.modules.keys())
        self.parser.add_option(
            "--output-module", metavar="MODULE",
            dest="OutputModule", default="portaudio",
            help="output modules: " + output_modules)
        self.parser.add_option(
            "--input", metavar="INPUT",
            dest="Input", default="0",
            help="input file or device")
        self.parser.add_option(
            "--output", metavar="OUTPUT",
            dest="Output", default="0",
            help="output file or device")
        self.parser.add_option(
            "--drumkit", metavar="FILENAME",
            dest="DrumKit", default="DrumKit.xml",
            help="drum kit file")
        self.parser.add_option(
            "--multi", action="store_true",
            dest="Multichannel", default=False,
            help="multichannel output")
        self.parser.add_option(
            "--low", action="store_true",
            dest="LowQuality", default=False,
            help="low quality")
        self.parser.add_option(
            "--fftsize", metavar="INT",
            dest="FFTSize", default=1024,
            help="FFT window size")
        self.parser.add_option(
            "--csound", metavar="OPTS",
            dest="Csound", default="",
            help="custom Csound options")
        self.parser.add_option(
            "--debug", action="store_true",
            dest="Debug", default=False,
            help="display debug info")


    def parse(self, arglist):

        # parse args
        options, args = self.parser.parse_args(arglist)
        return options.__dict__



def main():

    # hello
    print "csdrummer, Copyright (C) 2008 Schlagg"
    print "csdrummer comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it under certain conditions. Look at LICENSE.txt for details."
    # init
    opts = CommandLineParser().parse(sys.argv)
    engine = CsoundEngine(opts)
    engine.run()
    while engine.get_status() == 0:
        time.sleep(1)
    engine.halt()
    print "status:", engine.get_status()


if __name__ == "__main__":
    main()



trunk/src/csound.py
# csDrummer: Record a drum session in a virtual studio
# Copyright (C) 2008 Schlagg
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import string
import tempfile
import os
import threading
import logging

import csnd

import drumkit

class DrumkitScript:

    def __init__(self):

        self.text = ""

    def createInstrument(self):

        code = """
iInstrument pycall1i "createInstrument", 0
"""
        self.text += code


    def createTechnique(self):

        code = """
iTechnique pycall1i "createTechnique", iInstrument
"""
        self.text += code


    def createTarget(self):

        code = """
iTarget pycall1i "createTarget", iInstrument, iTechnique
"""
        self.text += code


    def createLayer(self):

        code = """
iLayer pycall1i "createLayer", iInstrument, iTechnique, iTarget
"""
        self.text += code


    def setTechniqueMIDI(self, midi):

        code = """
iMIDI = $MIDI
pycalli "setTechniqueMIDI", iInstrument, iTechnique, iMIDI
"""
        template = string.Template(code)
        subst = {}
        subst["MIDI"] = midi
        self.text += template.substitute(subst)


    def setTargetChannel(self, channel):

        code = """
iChannel = $Channel
pycalli "setTargetChannel", iInstrument, iTechnique, iTarget, iChannel
"""
        template = string.Template(code)
        subst = {}
        subst["Channel"] = channel
        self.text += template.substitute(subst)
        

    def setTargetGain(self, gain):

        code = """
iGain = $Gain
pycalli "setTargetGain", iInstrument, iTechnique, iTarget, iGain
"""
        template = string.Template(code)
        subst = {}
        subst["Gain"] = gain
        self.text += template.substitute(subst)


    def setTargetADSR(self, attack, decay, sustain, release):

        code = """
iAttack = $Attack
iDecay = $Decay
iSustain = $Sustain
iRelease = $Release
pycalli "setTargetADSR", iInstrument, iTechnique, iTarget, iAttack, iDecay, iSustain, iRelease
"""
        template = string.Template(code)
        subst = {}
        subst["Attack"] = attack
        subst["Decay"] = decay
        subst["Sustain"] = sustain
        subst["Release"] = release
        self.text += template.substitute(subst)
        
    
    def sortLayersByVel(self):

        code = """
pycalli "sortLayersByVel", iInstrument, iTechnique, iTarget
"""
        self.text += code        


    def setLayerVelocity(self, velocity):

        code = """
iVelocity = $Velocity
pycalli "setLayerVelocity", iInstrument, iTechnique, iTarget, iLayer, iVelocity
"""
        template = string.Template(code)
        subst = {}
        subst["Velocity"] = velocity
        self.text += template.substitute(subst)


    def setLayerGain(self, gain):

        code = """
iGain = $Gain
pycalli "setLayerGain", iInstrument, iTechnique, iTarget, iLayer, iGain
"""
        template = string.Template(code)
        subst = {}
        subst["Gain"] = gain
        self.text += template.substitute(subst)


    def setLayerSample(self, sample):

        code = """
iSample = $Sample
pycalli "setLayerSample", iInstrument, iTechnique, iTarget, iLayer, iSample
"""
        template = string.Template(code)
        subst = {}
        subst["Sample"] = sample
        self.text += template.substitute(subst)


    def setSampleData(self, sample, data):

        code = """
iSample = $Sample
SFile = "$Data"
iSampleData ftgen 0, 0, 0, -1, SFile, 0, 0, 1
pycalli "setSampleData", iSample, iSampleData
"""
        template = string.Template(code)
        subst = {}
        subst["Sample"] = sample
        subst["Data"] = data
        self.text += template.substitute(subst)
        

class DrumkitScriptProducer:

    def __init__(self, path, drumkit):
        
        self.__path = path
        self.__drumkit = drumkit
        self.script = DrumkitScript()
        self.__preprocessSamples()
        self.__preprocessChannels()
        self.__produceSamples()
        self.__produceDrumkit()
        

    def __produceLayer(self, layer):

        self.script.createLayer()
        self.script.setLayerVelocity(layer.getVelocity())
        self.script.setLayerGain(layer.getGain())
        self.script.setLayerSample(self.__samplesReplace[layer.getSample()])


    def __produceTarget(self, target):

        self.script.createTarget()
        channel = self.__channelsReplace[target.getChannel()]
        self.script.setTargetChannel(channel)
        self.script.setTargetGain(target.getGain())
        a, d, s, r = target.getADSR()
        self.script.setTargetADSR(a, d, s, r)
        for layer in target.getLayers():
            self.__produceLayer(layer)
        self.script.sortLayersByVel()


    def __produceTechnique(self, technique):

        self.script.createTechnique()
        for midi in technique.getMIDI():
            self.script.setTechniqueMIDI(midi)
        for target in technique.getTargets():
            self.__produceTarget(target)


    def __produceInstrument(self, instrument):

        self.script.createInstrument()
        for technique in instrument.getTechniques():
            self.__produceTechnique(technique)
        
    
    def __produceDrumkit(self):

        for instrument in self.__drumkit.getInstruments():
            self.__produceInstrument(instrument)
    

    def __produceSamples(self):
        
        for sample in self.__samplesData.iterkeys():
            sampleData = self.__samplesData[sample]
            sampleData = os.path.join(self.__path, sampleData)
            sampleData = os.path.normpath(sampleData)
            sampleData = os.path.normcase(sampleData)            
            self.script.setSampleData(sample, sampleData)
        
           
    def __preprocessSamples(self):
        
        samplesList = self.__drumkit.getSamples()
        samplesData = {}
        samplesReplace = {}
        sampleN = 0
        for sample in samplesList:
            samplesData[sampleN] = sample
            samplesReplace[sample] = sampleN
            sampleN += 1
        self.__samplesData = samplesData
        self.__samplesReplace = samplesReplace
        
    
    def __preprocessChannels(self):
        
        channelsList = self.__drumkit.getChannels()
        channelsReplace = {}
        channelN = 0
        for channel in channelsList:
            channelsReplace[channel] = channelN
            channelN += 1
        self.__channelsReplace = channelsReplace


def produceDrumkitScript(path, drumkit):
    
    producer = DrumkitScriptProducer(path, drumkit)
    return producer.script.text


inputModules = {
    "default": " ",
    "portmidi": " -+rtmidi=portmidi --midi-device=$Input",
    "alsa": " -+rtmidi=alsa --midi-device=$Input",
    "mme": " -+rtmidi=mme --midi-device=$Input",
    "virtual": " -+rtmidi=virtual --midi-device=0",
    "file": " -+rtmidi=null --midifile=$Input -T"
}


def produceInput(module, input):
    
    template = string.Template(inputModules[module])
    subst = {}
    subst["Input"] = input
    text = template.substitute(subst)
    return text
    

outputModules = {
        "default": " ",
        "portaudio": "-+rtaudio=portaudio --output=dac$Output",
        "alsa": "-+rtaudio=alsa --output=dac$Output",
        "jack": "-+rtaudio=jack --output=dac$Output",
        "mme": "-+rtaudio=mme --output=dac$Output",
        "coreaudio": "-+rtaudio=coreaudio --output=dac$Output",
        "file": "-+rtaudio=null --output=$Output"
    }


def produceOutput(module, output):
    
    template = string.Template(outputModules[module])
    subst = {}
    subst["Output"] = output
    text = template.substitute(subst)
    return text
    

class ScriptOptions:
    
    def __init__(self):
    
        # script parts
        self.SamplerCSD = "sampler.csd"
        self.SamplerPy = "sampler.py"
        # input/output
        self.InputModule = "default"
        self.Input = 0
        self.OutputModule = "default"
        self.Output = ""
        # drumkit
        self.Drumkit = ""
        # modes
        self.Multichannel = False
        self.LowQuality = False
        self.FFTSize = 1024
        # humanizer
        self.HumanFalloff = 1
        self.HumanTime = 0.1
        # Csound
        self.Csound = ""
        self.CsoundDefault = " -d"
        # Debug
        self.Debug = False
     

class ScriptProducer:
    

    def __init__(self, options):

        self.__options = options
        # locate modules dir
        self.__modulesPath = os.path.dirname(__file__)
        # parse drumkit
        self.__parser = drumkit.parseFile(self.__options.Drumkit)
        # configure
        self.__subst = {}
        self.__configModes()
        self.__configDrumkitLoader()
        self.__configHumanize()
        self.__configCommandLine()
        self.__configSampler()
        self.__configDebug()        
        # load template csd
        f = open(os.path.join(self.__modulesPath, self.__options.SamplerCSD), "r")
        template = string.Template(f.read())
        f.close()
        # make csd
        self.text = template.substitute(self.__subst)


    def __configModes(self):

        # Multichannel
        if self.__options.Multichannel:
            self.__subst["NumChannels"] = str(len(self.__parser.getChannels()))
        else:
            self.__subst["NumChannels"] = "1"
        # Quality
        if self.__options.LowQuality:
            self.__subst["LowQuality"] = "1"
        else:
            self.__subst["LowQuality"] = "0"
        # FFT Size
        self.__subst["FFTSize"] = self.__options.FFTSize
        
    
    def __configDrumkitLoader(self):
        
        path = os.path.dirname(self.__options.Drumkit)
        script = produceDrumkitScript(path, self.__parser)        
        self.__subst["DrumkitLoader"] = script
        
    
    def __configHumanize(self):
        
        # macro: HumanTime
        self.__subst["HumanTime"] = self.__options.HumanTime
        # macro: HumanFalloff
        self.__subst["HumanFalloff"] = self.__options.HumanFalloff
        
    
    def __configCommandLine(self):

        csoundOpts = ""
        # default opts
        csoundOpts += " " + self.__options.CsoundDefault
        # debug flag
##        if self.__options.Debug:
##            csoundOpts += " -v"
        # input/output
        csoundOpts += " " + produceInput(self.__options.InputModule, self.__options.Input)
        csoundOpts += " " + produceOutput(self.__options.OutputModule, self.__options.Output)
        # command-line options
        csoundOpts += " " + self.__options.Csound
        self.__subst["CsoundOptions"] = csoundOpts
        
    
    def __configDebug(self):
        
        if self.__options.Debug:
            self.__subst["Debug"] = "1"
        else:
            self.__subst["Debug"] = "0"
            
    
    def __configSampler(self):

        self.__subst["SamplerScript"] = os.path.join(self.__modulesPath, self.__options.SamplerPy)
        
        
def produceScript(options):
    
    producer = ScriptProducer(options)
    return producer.text
        

class CsoundCallback(csnd.CsoundCallbackWrapper):
    
    def __init__(self, csound):
        
        csnd.CsoundCallbackWrapper.__init__(self, csound)
        self.logger = logging.getLogger("Csound")
        self.messageBuffer = ""

    
    def MessageCallback(self, attr, msg):
        
        self.messageBuffer += msg
        lines = self.messageBuffer.splitlines(True)
        self.messageBuffer = ""
        for line in lines:
            if line.endswith("\n"):
                self.logger.debug(line.strip("\n"))
            else:
                self.messageBuffer += line

        

class CsoundEngine:

    def __init__(self, script):

        # init
        self.csound = csnd.Csound()
        self.callback = CsoundCallback(self.csound)
        self.callback.SetMessageCallback()
        self.status = 0
        self.time = 0
        self.__play = False
        self.__stop = False
        # create temporary CSD
        tempFilename = tempfile.mktemp(suffix=".csd")
        f = open(tempFilename, "w");
        f.write(script)
        f.close()
        # compile
        self.csound.Compile(tempFilename)
        # remove temporary
        os.remove(tempFilename)
        # create thread
        self.thread = threading.Thread(target=self.__threadMain)
        self.thread.start()
        

    def __threadMain(self):

        self.status = 0
        while self.status == 0:
            if self.__play:
                self.status = self.csound.PerformBuffer()
                self.time = self.csound.GetScoreTime()
            if self.__stop:
                self.csound.Stop()                
        self.csound.Cleanup()                
   
 
    def join(self):

        self.thread.join()


    def pause(self):

        self.__play = False
        
    
    def play(self):
        
        self.__play = True


    def stop(self):

        self.__stop = True
        self.thread.join()


    def getStatus(self):

        return self.status


    def getTime(self):

        return self.time
